### DISCUSSION TITLE
Issue with Accessing Optimized Data Block Addresses for FUXA Integration

### DISCUSSION THREAD

**Discussion by sephi251090** [frangoteam/FUXA#1171] [Q&A]
Hello,

I've encountered a challenge while trying to integrate Siemens PLC data with the FUXA SCADA interface. The issue arises with accessing variable addresses within optimized data blocks in TIA Portal.

For context, FUXA requires direct addressing to read and write to the PLC. However, when a data block (DB) is set as "optimized" in TIA Portal, it doesn't expose fixed addresses for its tags – instead, Siemens manages memory allocation automatically to enhance efficiency and performance. This presents a problem for SCADA systems that rely on direct addressing, as there are no static addresses available to be referenced.

The common workaround is to disable the optimization of the data blocks, thereby gaining access to static addresses. However, this approach negates the benefits of optimization such as reduced memory usage and potentially faster processing.

I'm looking for a solution that allows FUXA to access the necessary data without compromising on the optimization benefits within TIA Portal. Does FUXA support symbolic addressing that could resolve this issue? If not, what are the recommended practices for ensuring reliable communication between an optimized Siemens PLC and FUXA?

Any guidance on this would be greatly appreciated.

Thanks !

**Comment by unocelli** [comment]
Hi, FUXA is using [node-snap7](https://www.npmjs.com/package/node-snap7) I don't know if there is an equally good library

**Comment by oacalix** [comment]
Hello.

The problem of Tags in DBS that have an active access is that this memory area is different and conventional libraries such as Libnodave or Snap7 cannot access it. I have developed a driver to access this memory. If you like you can contact ocalix1995@gmail.com

**Comment by sephi251090** [comment]
hey, thanks, i send you an email

**Comment by martinjug** [comment]
Hi,

Did you check if this PLC has integrated OPC UA server? Many siemens PLCs have it, just need to enable it and you can communicate to FUXA.

Regards

**Reply by jbryan189** [reply]
I'd also suggest using OPC to access the S7. I've have great success using the OPC in Fuxa, as I struggled with the Optimized Blocks.
